frontend/lib/framework/error-classes-list.cpp

```diff
@@ -537,4 +537,138 @@ void ErrorProcDefExplicitAnonFormal::write(ErrorWriterBase& wr) const {
   wr.code(fn, {formal});
 }
 
+/* parser errors */
+
+void ErrorParsing::write(ErrorWriterBase& wr) const {
+  auto loc = std::get<const Location>(info);
+  auto errorMessage = std::get<std::string>(info);
+  assert(errorMessage.back() != '.' &&
+         "expected no period at end of ErrorParsing message");
+  errorMessage.push_back('.');
+  wr.heading(kind_, type_, loc, errorMessage);
+}
+
+void ErrorCannotAttachPragmas::write(ErrorWriterBase& wr) const {
+  auto loc = std::get<const Location>(info);
+  wr.heading(kind_, type_, loc, "cannot attach pragmas to this statement.");
+}
+
+void ErrorInvalidIndexExpr::write(ErrorWriterBase& wr) const {
+  auto loc = std::get<const Location>(info);
+  wr.heading(kind_, type_, loc, "invalid index expression.");
```

Comment:
Same thing here. Looking at the code, looks like we're trying to report having non-identifier expressions in loops. This can be worded as "Only identifiers, like `x`, or tuple expressions, like `(x,y)`, are allowed at this point in the loop".

It's a shame we don't have type information right now, otherwise we could even recommend what to write there.

In general, I see no reason to _not_ report `wr.code(loc)` in these.

Reply:
Just `wr.code(loc)` doesn't compile, it needs to be given the specific expression, but I can modify these errors to take the relevant `AstNode`.

Reply:
Ah, and the `AstNode`s don't have locations associated with them at this point, so as far as I can tell there isn't a way to print with `code` here.

Reply:
`wr.code` should in principle be usable for showing code via locations. I'll investigate why it's not working when I'm back if you don't get to it by then :)

Reply:
Just needed to fix up `locate`, which was missing a trivial overload for a location!

```diff
diff --git a/frontend/include/chpl/framework/ErrorWriter.h b/frontend/include/chpl/framework/ErrorWriter.h
index 29cfad0d90..f6b69aa01d 100644
--- a/frontend/include/chpl/framework/ErrorWriter.h
+++ b/frontend/include/chpl/framework/ErrorWriter.h
@@ -57,6 +57,9 @@ inline Location locate(Context* context, const ID& id) {
 inline Location locate(Context* context, const uast::AstNode* node) {
   return locate(context, node->id());
 }
+inline Location locate(Context* context, const Location& loc) {
+  return loc;
+}
 
 /// \cond DO_NOT_DOCUMENT
 /**
diff --git a/frontend/lib/framework/error-classes-list.cpp b/frontend/lib/framework/error-classes-list.cpp
index 99f9b64733..2f124cb342 100644
--- a/frontend/lib/framework/error-classes-list.cpp
+++ b/frontend/lib/framework/error-classes-list.cpp
@@ -693,8 +693,10 @@ void ErrorCommentEOF::write(ErrorWriterBase& wr) const {
   auto nestedLoc = std::get<2>(info);
   wr.heading(kind_, type_, loc, "end-of-file in comment.");
   wr.note(startLoc, "unterminated comment started here");
+  wr.code(startLoc);
   if (!nestedLoc.isEmpty()) {
     wr.note(nestedLoc, "nested comment started here");
+    wr.code(nestedLoc);
   }
 }
 ```

Reply:
You can just apply it here, since I really think we need code output for a lot of these error messages to look even better :)

Reply:
Yeah, I definitely want that functionality in before merging this, parse errors could really use it. I'll go ahead and add it and then add code to everything I can!